src/common/reactor.cc

```diff
@@ -0,0 +1,594 @@
+/* 
+   Mathieu Stefani, 15 juin 2016
+   
+   Implementation of the Reactor
+*/
+
+#include "reactor.h"
+
+namespace Aio {
+
+struct Reactor::Impl {
+
+    Impl(Reactor* reactor)
+        : reactor_(reactor)
+    { }
+
+    virtual ~Impl() = default;
+
+    virtual Reactor::Key addHandler(
+            const std::shared_ptr<Handler>& handler, bool setKey) = 0;
+
+    virtual std::vector<std::shared_ptr<Handler>> handlers(
+            const Reactor::Key& key) const = 0;
+
+    virtual void registerFd(
+            const Reactor::Key& key,
+            Fd fd,
+            Polling::NotifyOn interest,
+            Polling::Tag tag,
+            Polling::Mode mode = Polling::Mode::Level) = 0;
+
+    virtual void registerFdOneShot(
+            const Reactor::Key& key,
+            Fd fd,
+            Polling::NotifyOn interest,
+            Polling::Tag tag,
+            Polling::Mode mode = Polling::Mode::Level) = 0;
+
+    virtual void modifyFd(
+            const Reactor::Key& key,
+            Fd fd,
+            Polling::NotifyOn interest,
+            Polling::Tag tag,
+            Polling::Mode mode = Polling::Mode::Level) = 0;
+
+    virtual void runOnce() = 0;
+    virtual void run() = 0;
+
+    virtual void shutdown() = 0;
+
+    Reactor* reactor_;
+};
+
+/* Synchronous implementation of the reactor that polls in the context
+ * of the same thread
+ */
+struct SyncImpl : public Reactor::Impl {
+
+    SyncImpl(Reactor* reactor)
+        : Reactor::Impl(reactor)
+    {
+        shutdownFd.bind(poller);
+    }
+
+    Reactor::Key addHandler(
```

Comment:
TODO:

`addHandler` is currently not thread-safe, meaning that if the function is called from multiple threads, things will go boom boom. However, it whould be thread-safe and wait-free with respect to reads. To make it thread-safe, we could use a `mutex` to protect the writes and use some kind of atomic `shared_ptr` implementation to protect the reads